skbio/metadata/_interval.py

```diff
@@ -0,0 +1,465 @@
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from ._intersection import IntervalTree
+from skbio.util._decorator import experimental
+from itertools import chain
+import operator
+
+
+class Interval:
+    '''Stores the position and metadata of an interval.
+
+    This class supports the storing data for contiguous intervals
+    and non-contiguous intervals.
+
+    Parameters
+    ----------
+    interval_metadata : object
+        A reference to the `IntervalMetadata` object that this
+        interval is associated to.
+    intervals : iterable of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : iterable of tuple of bool
+        List of tuples, representing the openness of each interval.
+        If this isn't specified, then all of the boundaries are true.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+
+    See Also
+    --------
+    skbio.metadata.IntervalMetadata
+    '''
+    def __init__(self, interval_metadata, intervals,
+                 boundaries=None, metadata=None):
+
+        ivs = list(intervals)
+
+        # Intervals
+        self._interval_metadata = interval_metadata
+        # http://stackoverflow.com/a/6422754/1167475
+        # Used to sort boundaries later
+
+        indices = [i[0] for i in sorted(enumerate(ivs),
+                                        key=operator.itemgetter(1))]
+        self.intervals = sorted(ivs)
+
+        # Boundaries
+        if boundaries is not None:
+            bds = list(boundaries)
+            if len(bds) != len(ivs):
+                msg = ('The number of boundaries (%d) needs '
+                       'to be equal to the number of intervals (%d).')
+                raise ValueError(msg % (len(bds), len(ivs)))
+            self.boundaries = [bds[i] for i in indices]
+        else:
+            self.boundaries = [(True, True)] * len(ivs)
+
+        # Metadata
+        if metadata is not None:
+            self._metadata = metadata
+        else:
+            self._metadata = {}
+
+        if interval_metadata is not None:
+            self._add_interval()
+
+    def _add_interval(self):
+        """ Add an interval object to the IntervalTree within
+            IntervalMetadata."""
+        # Add directly to the tree.  So no need for _is_stale_tree
+        for loc in self.intervals:
+            start, end = loc
+            self._interval_metadata._intervals.add(start, end, self)
+        self._interval_metadata._metadata.append(self)
+
+    # This is required for creating unique sets of intervals
+    def _hash(self):
+        return hash(tuple(sorted(self.metadata.items()) +
```

Comment:
This hash function won't work because you may have unhashable items in metadata. For example:

```python
In [3]: md = {'foo': [1, 2, 3]}

In [4]: hash(tuple(md.items()))

TypeError: unhashable type: 'list'
```

In fact, the hashing isn't required here. We're going to note in ``IntervalMetadata.query`` how you can do this without a hash function.